backend/src/cache/deployer/deploy-cache-service.sh

```diff
@@ -38,3 +38,6 @@ echo "CA_BUNDLE patched successfully"
 # Create MutatingWebhookConfiguration
 cat ./cache-configmap-ca-bundle.yaml
 kubectl apply -f ./cache-configmap-ca-bundle.yaml --namespace "${NAMESPACE}"
+
+# Delete the cache deployer job
+kubectl delete job cache-deployer-job -n "${NAMESPACE}"
```

Comment:
Does this job need to get run for each upgrade?

Reply:
Will there be scenarios you want to provide upgrade support for this Job, e.g. you need to change the script?

Reply:
Not necessary. The job needs to be run only when cache-server namespace gets changed.

Reply:
I see. With current implementation, each time user upgrades kfp, this job will run one more time, it doesn't seem ideal, just double check, will extra job runs be idempotent?

Reply:
Yes, it is. With multiple deployer job run, the later will re-generate the certification and override the previous one. So does MutatingWebhookConfiguration.

Reply:
I think it's beneficial that the job runs on every update. There could be bug fixes or changes. An upgraded cluster should be equivalent to a newly-installed cluster.
Then there is a question what should the deployer script do.
Should it only deploy the configuration if it's missing or should it re-deploy?
I we plan to make any changes to the script of the manifests deployed by the script I'd favor re-deployment. Changes might also be hidden (for example a namespace change).

Reply:
If we use statefulset, it will also be rerun each upgrade because image changed

Reply:
As far as I know, both Deployment and StatefulSet are designed for services that should always be running (as indicated by the supported values of the `restartPolicy`). Whereas Job is specifically designed for tasks that go from start to end. See [Jobs - Run to Completion](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/).

>A Job creates one or more Pods and ensures that a specified number of them _successfully terminate_. As pods _successfully complete_, the Job tracks the _successful completions_. When a specified number of _successful completions_ is reached, the task (ie, Job) is complete.

Reply:
Now, I'm fine with both.

Current implementation using statefulset also make it long running (e.g. always try to update the configmap if it doesn't exist periodically, this may potentially be beneficial if some dependencies of this task isn't met in the cluster initially, but to be honest I don't know any real use cases now).

The only reason it was suggested was because Kubeflow was handling it this way. So I suggested @rui5i to ask about Kubeflow team for some reasoning.

Reply:
@IronPan @dushyanthsc  What do you think?

Reply:
FYI: People from Kubeflow replies:
"The expectation is if you run kubectl apply again this should be an idempotent action.
But if you use a job and it ran previously it will not run again even if thee spec changed
"Run once semantics" are also probably not what you want because infrastructure should be self-healing.
i.e. if the cert or whatever is deleted it should ideally get recreated at some point.
That said I think the new pattern with admission hooks is to use cert-manager to provision the certs"

Reply:
Can we make the deployer as part of the caching service? 

The caching service will have a sudo logic of following 
```
main(){
   bool exist = check_secret_exist();
   if(!exist){
       run_script_and_generate_tls_cert(); // store at /etc/webhook/certs
       create_secret();
   }
  else {
       read_from_secret(); 
  }
}
```

We have similar problem to solve for API server to load samples OAOO, and it's handled with the same pattern. 
https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/main.go#L167

This will avoid the upgradability issue all together.

Reply:
> Can we make the deployer as part of the caching service?
>We have similar problem to solve for API server to load samples

Thank you for the interesting suggestion.
This might be a bit harder in this case. The deployer needs different permissions and roles compared to the server. It also runs on the cluster level while the server is currently namespace-bound. There are also secrets that need to exist to be mounted to the server pod.

In the interest of time we'll probably leave deployer separate, but we'll explore the proposal to make the server self-deploying.

Thanks.